coveo-functools/coveo_functools/flex/deserializer.py

```diff
@@ -55,32 +61,91 @@ def convert_kwargs_for_unpacking(dirty_kwargs: Dict[str, Any], *, hint: MetaHint
         if arg_name not in mapped_kwargs:
             continue  # this may be ok, for instance if the target argument has a default
 
-        converted_kwargs[arg_name] = deserialize(mapped_kwargs[arg_name], hint=arg_hint)
+        converted_kwargs[arg_name] = deserialize(
+            mapped_kwargs[arg_name], hint=arg_hint, errors=errors
+        )
 
     return converted_kwargs
 
 
 @overload
-def deserialize(value: Any, *, hint: Type[T]) -> T:
+def deserialize(
+    value: Any,
+    *,
+    hint: Type[T],
+    errors: ErrorBehavior = "deprecated",
+) -> T:
     ...
 
 
 @overload
-def deserialize(value: Any, *, hint: T) -> T:
+def deserialize(
+    value: Any,
+    *,
+    hint: T,
+    errors: ErrorBehavior = "deprecated",
+) -> T:
     """This overload tricks mypy in passing typing annotations as types, such as List[str]."""
 
 
-def deserialize(value: Any, *, hint: Union[T, Type[T]]) -> T:
+def deserialize(
+    value: Any,
+    *,
+    hint: Union[T, Type[T]],
+    errors: ErrorBehavior = "deprecated",
+) -> T:
     """
     Deserializes a value based on the provided type hint:
 
     - If value is None, we return None. We don't validate the hint.
     - If hint is a builtin type, we blindly return the value we were given. We don't validate the value.
     - If hint is a custom class, the value must be a dictionary to "flex" into the class.
     - Hint may be a `Union[T, List[T]]`, in which case the value must be a dict or a list of them.
+
+    The `errors` argument controls the behavior when a value cannot be deserialized into the hint or annotation:
+        - 'ignore': the value is used as-is, without deserialization. errors will be logged.
+        - 'silent': behave like 'ignore' but don't log errors  # yolo
+        - 'raise': raise a PayloadMismatch exception.
+        - 'deprecated': (default) different situations yield different behaviors:
+            - type mismatch between value and hint (e.g.: cannot use a list to create a dict) would behave like 'ignore'
+            - value mismatch (e.g.: dict missing some arguments vs a class's __init__) would raise TypeError
+
+    Note that the `deprecated` option is currently the default, for legacy reasons. It will be removed because
+    it created an inconsistent behavior (see the examples below).
+
+    example #1: `range` requires the `stop` argument, which is missing:
+
+        >>> deserialize({}, hint=range, errors='raise')  # doctest: +IGNORE_EXCEPTION_DETAIL
+        Traceback (most recent call last):
+        TypeError: range expected 1 argument, got 0
+        >>> deserialize({}, hint=range, errors='deprecated')  # doctest: +IGNORE_EXCEPTION_DETAIL
+        Traceback (most recent call last):
+        TypeError: range expected 1 argument, got 0

```

Comment:
I think there might be a copy-paste error here. Both cases look very similar.

Reply:
the first one has raise, the second has deprecated. The behavior is the same. However in the other example, the behavior of "deprecated" is different. That's what I tried to illustrate here. The middle choice (deprecated) is not consistent, since it would raise in one situation and ignore in the other.